cmds/cat/cat.go

```diff
@@ -16,42 +16,38 @@ package main
 
 import (
 	"flag"
-	"fmt"
 	"io"
+	"log"
 	"os"
 )
 
 var (
 	_ = flag.Bool("u", false, "ignored")
 )
 
-func cat(writer io.Writer, files []string) error {
-	for _, name := range files {
-		f, err := os.Open(name)
-		if err != nil {
-			return err
-		}
-
-		_, err = io.Copy(writer, f)
-		if err != nil {
-			return err
-		}
-		f.Close()
+func cat(w io.Writer, file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		return err
 	}
+	defer f.Close()
```

Comment:
It's actually been recommended recently that you avoid defer on closes unless you absolutely need it because there are no guarantees about when it will run. I've actually been burned by 'deferred defers' as a result. 
Since you return after the copy I'd suggest doing the close then.

Reply:
What? It should be guaranteed to run before this function returns. Deferred defers are not a thing, they must be unwound on the stack.

In fact, I thought it was recommended that you defer on closes, because you have absolutely no guarantees with finalizers.

Reply:
This is why I changed this function to only open one file and output that -- we'll be guaranteed one file is closed before the next one is opened.